app/commands/user/challenges/featured_exercises_progress_12_in_23.rb

```diff
@@ -0,0 +1,49 @@
+class User::Challenges::FeaturedExercisesProgress12In23
+  include Mandate
+
+  initialize_with :user
+
+  def call
+    featured_exercises.filter_map do |exercise_slug, track_slugs|
+      next unless solutions.key?(exercise_slug)
+
+      solved_in_featured_tracks = solutions[exercise_slug].select { |track_slug| track_slugs.include?(track_slug) }
+
+      solved_in_23 = solved_in_featured_tracks.select { |_, year| year == 2023 }
+      next [solved_in_23.keys.first, exercise_slug] if solved_in_23.any?
+
+      solved_before_23 = solved_in_featured_tracks.select { |_, year| year < 2023 }
+      next [solved_before_23.keys.first, exercise_slug] if solved_before_23.size == track_slugs.size
+    end
+  end
+
+  private
+  memoize
+  def solutions
+    user.solutions.published.
```

Comment:
I was slightly confused about whether the solution has to be published or not for it to count.

The "The year-long #12in23 badge" section has this description:

> At the end of the year we'll be awarding an exclusive badge to anyone who completes all the featured exercises in the correct paradigms. You can complete the exercise any time during 2023. Each valid exercise will be shown below

No mention of published there, but the exercises it lists are checked via published (which I copied).